File: evaluation/S0/python/Easy/334-special-positions-in-a-binary-matrix.py
# Source: https://leetcode.com/problems/special-positions-in-a-binary-matrix/   |   Difficulty: Easy
#
# Problem Description:
# Given an m x n binary matrix mat, return the number of special positions in mat.
#
# A position (i, j) is called special if mat[i][j] == 1 and all other elements in row i and column j are 0 (rows and columns are 0-indexed).
#
# Example:
# Input: mat = [[1,0,0],[0,0,1],[1,0,0]]
# Output: 1
# Explanation: (1, 2) is a special position because mat[1][2] == 1 and all other elements in row 1 and column 2 are 0.
#
# Constraints:
# m == mat.length
# 	n == mat[i].length
# 	1 <= m, n <= 100
# 	mat[i][j] is either 0 or 1.
#

class Solution(object):
    def numSpecial(self, mat):
        """
        :type mat: List[List[int]]
        :rtype: int
        """
        row_counts = [0] * len(mat)
        col_counts = [0] * len(mat[0])
        
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                if mat[i][j] == 1:
                    row_counts[i] += 1
                    col_counts[j] += 1
        
        special_positions = 0
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                if mat[i][j] == 1 and row_counts[i] == 1 and col_counts[j] == 1:
                    special_positions += 1
        
        return special_positions

File: evaluation/S0/java/Easy/334-special-positions-in-a-binary-matrix.java
// Source: https://leetcode.com/problems/special-positions-in-a-binary-matrix/   |   Difficulty: Easy
//
// Problem Description:
// Given an m x n binary matrix mat, return the number of special positions in mat.
//
// A position (i, j) is called special if mat[i][j] == 1 and all other elements in row i and column j are 0 (rows and columns are 0-indexed).
//
// Example:
// Input: mat = [[1,0,0],[0,0,1],[1,0,0]]
// Output: 1
// Explanation: (1, 2) is a special position because mat[1][2] == 1 and all other elements in row 1 and column 2 are 0.
//
// Constraints:
// m == mat.length
// 	n == mat[i].length
// 	1 <= m, n <= 100
// 	mat[i][j] is either 0 or 1.
//

class Solution {
    public int numSpecial(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;

        int[] rowCount = new int[m];
        int[] colCount = new int[n];

        // Count the number of 1's in each row and column
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 1) {
                    rowCount[i]++;
                    colCount[j]++;
                }
            }
        }

        int specialPositions = 0;

        // Identify special positions
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 1 && rowCount[i] == 1 && colCount[j] == 1) {
                    specialPositions++;
                }
            }
        }

        return specialPositions;            
    }
}
